File: binary_search/binary_search.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

template <typename T>
int binary_search(vector<T>& arr, int start, int end, int target)
{
    int mid = (start + end) / 2;
    int target_idx;

    if(arr[mid] < target)
    {
        binary_search(arr, start, mid, target);
    }

    else if (arr[mid] > target)
    {
        binary_search(arr, mid+1, end, target);
    }
    else if (arr[mid] == target)
    {
        return mid;
    }
}

File: binary_search/binary_search.py
def binary_search(arr, start, end, target):
    if start > end:
        return None
    
    mid = (start + end) // 2

    if arr[mid] == target:
        return mid
    elif arr[mid] > target:
        return binary_search(arr, start, mid - 1, target)
    else:
        return binary_search(arr, start, mid + 1, target)
    
def binary_search(arr, start, end, target):
    while start <= end:
        mid = (start + end) // 2
        if arr[mid] == target:
            return mid
        elif arr[mid] > target:
            end = mid - 1
        else:
            start = mid + 1
    return None
